### ISSUE TITLE
Allow for configuration which provider a function should be deployed to

### ISSUE THREAD

**Issue by flomotlik** [serverless/serverless#1328]
To allow multiple providers we need to implement a way for functions to only be deployed to specific providers.

Regarding selecting which functions should be deployed to which provider I propose the following:

A provider configuration setting on the highest level of the serverless.yml file that defines which provider all functions should be deployed to:

```
service: first_service
plugins:
  - additional_plugin
  - another_plugin
provider: aws
```

You can also decide for specific functions to be deployed to a different provider

```
functions:
  hello:
    handler: handler.hello
    provider: azure
```

Function specific configuration has precedence over service wide configuration.

We might allow to implement configuration through command line options as well which could overwrite all the different options in the config file, but not sure if we should do that.

To load only the functions that a specific provider should be deploying we can implement an additional function:

```
  getAllFunctionsForProvider(provider) {
    //Select functions according to service and function based config
  }
```

**Comment by pmuens** [comment]
Good ideas.

What if I want to deploy one service to different providers (don't know why I want to do this, but could be a possible use case). Is there a way that the `provider` property can be an array?

@eahefnawy and I were discussing that recently and @eahefnawy came up with the idea of having that as an option. So you could say `serverless deploy --stage dev --region us-east-1 --provider aws`. This would make the need to always modify the `serverless.yaml` obsolete.

We can implement this by adding a check inside the corresponding plugin like this:
```
if (this.options.provider !== 'aws') {
  return;
} 
```
or something similar.

This would make the whole process also flexible as custom providers can be integrated quite easily (they just need the check for the provider inside their plugin) and we don't have to add the provider support in our core framework.

Another idea would be to add the check inside the `PluginManager` for example. But this would also mean that we have to introduce each new provider to our core system.

**Comment by flomotlik** [comment]
Imho we have a couple of different use cases we need to make work:

1. Deploy all functions to the same provider
2. Deploy some functions to one provider, others to another provider
3. Deploy functions to several providers at once
4. A mix of 2 & 3

So imho setting the provider as an array can make sense (although I would prefer being able to set it as string or array so when you only deploy to one provider it makes the file shorter)

Imho plugins should have the option to define that they only work for specific providers and if those providers aren't set in the current service they simply shouldn't be used for the current command.

Otherwise you would have to implement the provider check in every hook (or at least in the constructor and then not add the hooks). Which can make sense for some plugins, but it shouldn't be necessary by default imho.

I do think the command line argument is a nice feature as it allows to overwrite things easily, but imho its so early in our multi-provider support that its not really possible to deploy the same function to multiple providers anyway, so I don't think people will be really using that feature. Once we see a need for it we can definitely implement it.

**Comment by pmuens** [comment]
Ok. That makes sense.
Yes, I also don't 100% align with the options approach but putting it inside of `serverless.yaml` feels also not correct as this file will be committed to source control.

However what about this approach:
We define a `deployment` property inside of `serverless.yaml` where everything deployment related lives (e.g. provider(s), stage, region). The values there are used by default but the user can overwrite them by passing in the corresponding options.

This way I can cd into the service folder and simply run `serverless deploy` and it gets deployed to the provider, stage and region defined inside the `serverless.yaml` file. Or I can run `serverless deploy --provider aws` to overwrite the provider property.

If multiple options are defined and I have not passed in an option I have to select them in the CLI (like in v0).

Here's how such a section in `serverless.yaml` might look like:
```
deployment:
  providers:
    - aws
    - azure
  stages:
    - dev
    - production
  regions:
    - us-east-1
```

**Comment by eahefnawy** [comment]
defining stages and regions in serverless.yaml is dangerous imo ... users could accidently deploy a function to production very very easily this way. I think when it comes to choosing stages (and maybe regions too) it should be completely explicit.

However, @flomotlik mentioned a while back adding config for a **default stage and region**. I really like that! users could just have:

```yml
service:
defaults:
    stage: dev
    region: us-east-1
functions:
    hello:
events:
resources:
```

During development, the user will just keep on running `sls deploy` and that's it! when it comes to deploying to production, he has to specifically provide stage/ region: `sls deploy --stage prod --region us-east-1

**Comment by eahefnawy** [comment]
When I suggested passing the provider as an option, it was based on our plans to deploy specific functions via the command line via `sls deploy --function functionName --provider aws` ... however now that we deploy the entire service, passing an option would make that provider option tied to the entire service and all its functions. **So now I wouldn't go with `--provider` option.**

One other thing that I realized that makes it even more complex, is the fact that functions contain events that are provider specific. So that means **each function can't be tied to a single provider** as it depends on the events. Maybe we can check the events that the function has and deploy to all required providers automatically? hmmm but what if a user just want to deploy a function without any events.

I mean what would happen if we have the following config:

```yml
functions:
     hello:
           provider: aws
           events:
               azure_http:    # Oops, there's no function in azure! only in aws
```
This will require super strong validation by comparing each event provider with the function provider. A lot of conflicts might happen and the logic could be hairy.

Generally, I hardly doubt that users will have a single function that handles events from multiple providers. Specially that events from different providers will have different payload/input, even for similar events (ie. S3 bucket and azure bucket events will probably be completely different JSON data). So I dont think users will use the exact same function logic to handle both payloads. This calls for two different functions imo.

So with that in mind, maybe we should have the provider config to be **only tied to the function**, and as such, all its events will be tied to that provider. That touches on our event syntax discussion as we wont need provider nesting or prefixes for events at all. Events will just inherit function provider config:

```yml
service:
functions:
    function1:
        provider: aws
        events:
            s3:
    function2:
        provider: azure
        events:
           http:  # this is obviously an azure http cause the function will be deployed to azure
```

Thoughts...?

**Comment by ac360** [comment]
Really good discussion here.  Great job @flo summarizing the use-cases too.  That should always be front and center.  

After debating `serverless.yml` for the past few days, and reading the good ideas in here, here's a possible elegant solution…

I think the majority of our syntax problems come from trying to mix multiple providers into a single service (`serverless.yml` file).  This is forcing excessive and repetitive use of provider keys in our `functions`, `events` and `resources` properties, resulting in an ugly and impractical `serverless.yml` file, that is hard to parse.

Instead, what if we limited each service to a single provider, but then offer a way to nest services within each other?

**Here is an example of a multi-cloud service:**

```
serverless.yml
aws
  serverless.yml
azure
  serverless.yml
```

**./serverless.yml**
This file has no functions.  It’s just there to declare a service, and declare that any child services belong to it.  In `deploy` we'll have to create an initial step that checks for child services.  If any are found, those are deployed first.  This should also remove the need for a `--provider` flag and provider detection logic.
```
service: dataPipeline
```
**./aws/serverless.yml**
This just looks like @flomotlik's original example mixed w/ some of your other ideas, with provider at the root.
```
service: awsDataProcesses
provider: aws
functions:
  create:
    memorySize: 1024
  events:
    - s3:
    - scheduled:
    - http:
```

Nesting services is part of SOA.  There is a lot of precedent here.  `deploy` should probably do shallow searches (1-level deep) for the sake of simplicity, and since the only point of nesting will be to create multi-cloud services.

**Comment by pmuens** [comment]
Thinking through this I can say that I like the idea of having one provider per `serverless.yaml` file.

I also think that we should try to reduce options as much as possible and a single `serverless deploy´ should be enough to (re)deploy the whole service (no need to pass in options at all).
I also like the idea of defining a default and the overwrite it with options.

I would not start with a complex "Provider inheritance model" and just say "One provider per service" as @ac360 has mentioned.

One thing we need to keep in mind is that we might want to make the extensibility of the framework as easy as possible. I think it would be a downside if people write integrations for other providers and need to go through a PR to get support for it into the framework (I mean that the framework should not include hard coded provider specific code).

Another thing I have to note is that `serverless.yaml` is a file which will be tracked by source control. This might conflict with the idea of having the stage and region in the file (I think it's ok to have it there).

Here's a reference to issue https://github.com/serverless/serverless/issues/1324 which is related to this one.

**Comment by flomotlik** [comment]
I agree. For the most part you don't really want to deploy services to various providers anyway and trying to put that in from the beginning is more pain than its worth now.

Plugins can then simply set a provider attribute internally (which could be a list) of all the different providers they support and the PluginManager only uses the ones that match the currently configured provider.

This also makes it easy for the create plugin as it can take a ```--provider``` option and then simply use a different default template depending on the provider.

I wouldn't give the option of using a --provider cli option now to set a different one from the one in the config file. Imho the config files are so different that it doesn't really work anyway to simply switch providers.

So yeah I agree, @eahefnawy @pmuens I think you guys are already booked for today with finishing up APIG and Documentation/Cleanup, but lets discuss monday and make this new setup the plan for next week imho.

**Comment by pmuens** [comment]
Further discussion and implementation details in #1340
